lib/AST/LayoutConstraint.cpp

```diff
@@ -143,53 +140,142 @@ void *LayoutConstraintInfo::operator new(size_t bytes, const ASTContext &ctx,
+     E(/* Class */ Class), E(/* NativeClass */ NativeClass),
+     E(/* RefCountedObject*/ RefCountedObject),
+     E(/* NativeRefCountedObject */ NativeRefCountedObject)},
+
+    // Initiaze the row for TrivialOfExactSize.
+    {E(/* UnknownLayout */ TrivialOfExactSize),
+     E(/* TrivialOfExactSize */ TrivialOfExactSize), MERGE_CONFLICT,
+     E(/* Trivial */ TrivialOfExactSize), MERGE_CONFLICT, MERGE_CONFLICT,
+     MERGE_CONFLICT, MERGE_CONFLICT},
+
+    // Initiaze the row for TrivialOfAtMostSize.
+    {E(/* UnknownLayout */ TrivialOfAtMostSize), MERGE_CONFLICT,
+     E(/* TrivialOfAtMostSize */ TrivialOfAtMostSize),
+     E(/* Trivial */ TrivialOfAtMostSize), MERGE_CONFLICT, MERGE_CONFLICT,
+     MERGE_CONFLICT, MERGE_CONFLICT},
+
+    // Initiaze the row for Trivial.
+    {E(/* UnknownLayout */ Trivial),
+     E(/* TrivialOfExactSize */ TrivialOfExactSize),
+     E(/* TrivialOfAtMostSize */ TrivialOfAtMostSize), E(/* Trivial */ Trivial),
+     MERGE_CONFLICT, MERGE_CONFLICT, MERGE_CONFLICT, MERGE_CONFLICT},
+
+    // Initiaze the row for Class.
+    {E(/* UnknownLayout*/ Class), MERGE_CONFLICT, MERGE_CONFLICT,
+     MERGE_CONFLICT, E(/* Class */ Class), E(/* NativeClass */ NativeClass),
+     E(/* RefCountedObject */ Class),
+     E(/* NativeRefCountedObject */ NativeClass)},
+
+    // Initiaze the row for NativeClass.
+    {E(/* UnknownLayout */ NativeClass), MERGE_CONFLICT, MERGE_CONFLICT,
+     MERGE_CONFLICT, E(/* Class */ NativeClass),
+     E(/* NativeClass */ NativeClass), E(/* RefCountedObject */ NativeClass),
+     E(/* NativeRefCountedObject */ NativeClass)},
+
+    // Initiaze the row for RefCountedObject.
+    {E(/* UnknownLayout */ RefCountedObject), MERGE_CONFLICT, MERGE_CONFLICT,
+     MERGE_CONFLICT, E(/* Class */ Class), E(/* NativeClass */ NativeClass),
+     E(/* RefCountedObject */ RefCountedObject),
+     E(/* NativeRefCountedObject */ NativeRefCountedObject)},
+
+    // Initiaze the row for NativeRefCountedObject.
+    {E(/* UnknownLayout */ NativeRefCountedObject), MERGE_CONFLICT,
+     MERGE_CONFLICT, MERGE_CONFLICT, E(/* Class */ NativeClass),
+     E(/* NativeClass */ NativeClass),
+     E(/* RefCountedObject */ NativeRefCountedObject),
+     E(/* NativeRefCountedObject*/ NativeRefCountedObject)},
+};
+
+#undef E
+
+LayoutConstraint
+LayoutConstraint::merge(LayoutConstraint Other) {
+  auto Self = *this;
+
+  // If both constraints are the same, they are always equal.
+  if (Self == Other)
+    return Self;
+
+  // TrivialOfExactSize and TrivialOfAtMostSize are a special case,
+  // because not only their kind, but their parameters need to be compared as
+  // well.
+  if (Self->isKnownSizeTrivial() && Other->isKnownSizeTrivial()) {
+    // If we got here, it means that the Self == Other check above has failed.
+    // And that could only happen if either the kinds are different or
+    // size/alignment parameters are different.
+    return LayoutConstraint::getUnknownLayout();
```

Comment:
Okay, it probably doesn't matter, but one could merge exact-size(N) and at-most-size(N) into exact-size(N).